auslib/db.py

```diff
@@ -1756,21 +1756,40 @@ def __init__(self, db, metadata, dialect):
         AUSTable.__init__(self, db, dialect, scheduled_changes=True, scheduled_changes_kwargs={"conditions": ["time"]})
 
     def getPotentialRequiredSignoffs(self, affected_rows, transaction=None):
-        potential_required_signoffs = []
+        potential_required_signoffs = {}
+        rows = []
         for row in affected_rows:
             if not row:
                 continue
-            # Releases do not affect live updates on their own, only the
-            # product+channel combinations specified in Rules that point
-            # to them. We need to find these Rules, and then return _their_
-            # Required Signoffs.
-            info = self.getReleaseInfo(name=row["name"], transaction=transaction)
-            if info:
-                info = info[0]
-                for rule_id in info["rule_ids"]:
-                    rule = self.db.rules.select(where=[self.db.rules.rule_id == rule_id], transaction=transaction)[0]
-                    potential_required_signoffs.extend(self.db.rules.getPotentialRequiredSignoffs([rule], transaction=transaction))
-        return potential_required_signoffs
+            rows.append(row)
+        info = self.getReleaseInfo(names=[row['name'] for row in rows], transaction=transaction)
+        # Releases do not affect live updates on their own, only the
+        # product+channel combinations specified in Rules that point
+        # to them. We need to find these Rules, and then return _their_
+        # Required Signoffs.
+        if info:
+            # get all rules as a union
+            q_rules = [self.db.rules.rule_id.in_(tuple([rule_id for row in info for rule_id in row['rule_ids']]))]
+            all_rules = self.db.rules.select(where=q_rules, transaction=transaction)
+
+            # map rules according to rule_id
+            rules_map = {rule['rule_id']: rule for rule in all_rules}
+
+            rs_cache = {}
+
+            for row in info:
+                rs = []
+                potential_required_signoffs[row['name']] = []
+                for rule_id in row["rule_ids"]:
+                    rule = rules_map[rule_id]
+                    if rule_id in rs_cache:
+                        _rs = rs_cache[rule_id]
+                    else:
+                        _rs = self.db.rules.getPotentialRequiredSignoffs([rule], transaction=transaction)
+                        rs_cache[rule_id] = _rs
+                    rs.extend(_rs)
+                potential_required_signoffs[row['name']] = rs
+        return next(iter(potential_required_signoffs.values())) if len(potential_required_signoffs) is 1 else potential_required_signoffs
```

Comment:
It just clicked with me that we now have two different possible return types for this. One is the list, like all of the other `getPotentialRequiredSignoffs' functions, and the other is a dictionary. This kind of thing makes the code difficult to reason about - if there's no way to improve the query time without changing the return type, this new code is better off in another function. Not needing to support the list return type might let you simplify things a bit too.

Sorry for not catching this part earlier.

Reply:
@mozbhearsum No problem.
This is the case coz `getPotentialRequiredSignoffs` is expected to return an array when getting `releaseScheduledChanges` https://github.com/mozilla/balrog/blob/8da414c3762e0e5e87cbe041d3400de76619bd7b/auslib/web/admin/views/scheduled_changes.py#L62 . We could perhaps change this to use dictionaries so we make our return type a dictionary, or create another function. What do you think would be a better approach?

Reply:
It's that, as well as numerous invocations in db.py (like https://github.com/mozilla/balrog/blob/master/auslib/db.py#L1142). Updating them to support the dict return format is a really great idea - let's do that!

Reply:
Great! will do that then. Thanks